### PR TITLE
feat: implement add/remove controller and full update of a did doc

### PR DESCRIPTION
**Pull request by noandrea**
The new implementation allows to fully update a DID document
and to manipulate the controller section with dedicated functions

Another breaking change is relative to the chain events

### CODE REVIEW

File: x/did/keeper/msg_server.go [hunk 1]
```diff
@@ -368,3 +233,58 @@ func updateDidMetadata(keeper *Keeper, ctx sdk.Context, did string) (err error)
...
+	relationships []string
+}
+
+func newConstraints(relationships ...string) Constraints {
+	return Constraints{
+		relationships: relationships,
+	}
+}
+
+func executeOnDidWithRelationships(goCtx context.Context, k *Keeper, constraints Constraints, did, signer string, update func(document *types.DidDocument) error) (err error) {
```
Comment (PaddyMc) [inline x/did/keeper/msg_server.go:249]: This is awesome!!!! :mage: 

:heart:

**Review by PaddyMc — APPROVED** [review]
Great work with the PR, some small suggestions, but feel free to merge and we can colab on the extra bits to do :smile:

File: x/did/keeper/msg_server.go [hunk 2]
```diff
@@ -313,48 +173,53 @@ func (k msgServer) SetVerificationRelationships(
...
-	k.Keeper.SetDidDocument(ctx, []byte(msg.Id), didDoc)
-	k.Logger(ctx).Info("Set verification relationship from did document for", "did", msg.Id, "controller", msg.Signer)
+	return &types.MsgAddControllerResponse{}, nil
+}
 
-	// update the Metadata
-	if err := updateDidMetadata(&k.Keeper, ctx, didDoc.Id); err != nil {
-		k.Logger(ctx).Error("request to update did document metadata failed", "error", err.Error(), "did", didDoc.Id)
+// DeleteController remove an existing controller from a DID document
+func (k msgServer) DeleteController(
```
Comment (PaddyMc) [inline x/did/keeper/msg_server.go:209]: Do we need a cli command that calls this function in the msg_server?

File: x/did/client/cli/tx.go [hunk 3]
```diff
@@ -32,7 +32,6 @@ func GetTxCmd() *cobra.Command {
 		NewAddServiceCmd(),
 		NewRevokeVerificationCmd(),
 		NewDeleteServiceCmd(),
```
[the remark this answers is not in the document]
Comment (noandrea) [reply x/did/client/cli/tx.go:34]: whoopsy, thanks :+1:

File: x/did/keeper/msg_server.go (hunk 2 above)
Comment (noandrea) [reply x/did/keeper/msg_server.go:211]: yes we do!

**Review by PaddyMc — APPROVED** [review]

**Pull request merged**